### PR TITLE
feat: allow passing in user-defined context to Argo client

### PR DESCRIPTION
**Pull request by karenin-overseas**
Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Hello, this PR adds a `Context` field to the Argo Client opts so a user-defined context can be used in place of `context.Background()`. Basically go programs using the API Client might want the argo client to respect ex. context cancellation on a parent context.

I tested by running `make cli` and editing the CLI to use a custom context.

With thanks to the reviewers :)

### CODE REVIEW

File: pkg/apiclient/apiclient.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ type Client interface {
 
 type Opts struct {
 	ArgoServerOpts ArgoServerOpts
+	Context *context.Context
```
Comment (alexec) [inline pkg/apiclient/apiclient.go:30]: should not be pointer type

File: pkg/apiclient/argo-kube-client.go [hunk 2]
```diff
@@ -38,7 +38,7 @@ type argoKubeClient struct {
 
 var _ Client = &argoKubeClient{}
 
-func newArgoKubeClient(clientConfig clientcmd.ClientConfig, instanceIDService instanceid.Service) (context.Context, Client, error) {
+func newArgoKubeClient(clientConfig clientcmd.ClientConfig, instanceIDService instanceid.Service, baseContext *context.Context) (context.Context, Client, error) {
```
Comment (alexec) [inline pkg/apiclient/argo-kube-client.go:41]: context should be first arg by convention?

File: pkg/apiclient/apiclient.go (hunk 1 above)
Comment (alexec) [reply pkg/apiclient/apiclient.go:30]: if you add a `GetContext` func, then if `Context` is not set, you could just return `context.Background()`

**Review by alexec — CHANGES_REQUESTED** [review]
Great stuff. Good idea! Some comments...

**Comment by sarabala1979** [comment]
@karenin-overseas are you planning to address the review comment and get ready for review?

**Comment by karenin-overseas** [comment]
> @karenin-overseas are you planning to address the review comment and get ready for review?

Hi @sarabala1979 yes! I'll comment when it's ready for review.

**Comment by alexec** [comment]
Defunct: #6667

**Pull request closed without merging**